tls/extensions/s2n_server_psk.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <sys/param.h>
+#include <stdint.h>
+
+#include "crypto/s2n_hash.h"
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls_parameters.h"
+#include "tls/extensions/s2n_server_psk.h"
+
+#include "utils/s2n_bitmap.h"
+#include "utils/s2n_safety.h"
+
+#define CHOSEN_PSK_WIRE_INDEX_SIZE sizeof(uint8_t)
+
+static bool s2n_server_psk_should_send(struct s2n_connection *conn);
+static int s2n_server_psk_send(struct s2n_connection *conn, struct s2n_stuffer *out);
+static int s2n_server_psk_recv(struct s2n_connection *conn, struct s2n_stuffer *extension);
+
+const s2n_extension_type s2n_server_psk_extension = {
+    .iana_value = TLS_EXTENSION_PRE_SHARED_KEY,
+    .is_response = true,
+    .send = s2n_server_psk_send,
+    .recv = s2n_server_psk_recv,
+    .should_send = s2n_server_psk_should_send,
+    .if_missing = s2n_extension_noop_if_missing,
+};
+
+static bool s2n_server_psk_should_send(struct s2n_connection *conn)
+{
+    /* Only send a server pre_shared_key extension if a chosen PSK is set on the connection */
+    return conn && s2n_connection_get_protocol_version(conn) >= S2N_TLS13
+            && conn->psk_params.chosen_psk;
+}
+
+static int s2n_server_psk_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+
+    struct s2n_psk_parameters psk_params = conn->psk_params;
+
+    uint8_t choosen_psk_index = psk_params.chosen_psk_wire_index;
+    S2N_ERROR_IF(choosen_psk_index >= psk_params.psk_list.len, S2N_ERR_INVALID_ARGUMENT);
```

Comment:
This assignment is unnecessary, and you're losing some information by changing the name.

"chosen_psk_wire_index" is an index into the list of PSKs sent by the client / received over the wire, NOT the list stored locally. So you can't check its range within the local list.

Make sure you test a couple cases of the client PSK list not matching the server PSK list so you'll catch mistakes like this.

Reply:
In this function  `s2n_server_psk_send` the assumption is that a valid `choosen_psk_index` is already set by the server. The following validation MUST be done on the client pre_shared key extension during the `s2n_client_psk_recv` call. 

```
From RFC: https://tools.ietf.org/html/rfc8446#section-4.2.11
Clients MUST verify that the server's selected_identity is within the range supplied by the client
```

Reply:
In this function  `s2n_server_psk_send` the assumption is that a valid `chosen_psk_wire_index ` is already set by the server. The following validation MUST be done on the client pre_shared key extension during the `s2n_client_psk_recv` call. 

```
From RFC: https://tools.ietf.org/html/rfc8446#section-4.2.11
Clients MUST verify that the server's selected_identity is within the range supplied by the client
```

Reply:
In this function  `s2n_server_psk_send` the assumption is that a valid `chosen_psk_wire_index ` is already set by the server. The following validation MUST be done on the client pre_shared key extension during the `s2n_client_psk_recv` call which we have unit tests for on the client side psk extension already and adding tests for mismatch would be redundant to those tests. 

```
From RFC: https://tools.ietf.org/html/rfc8446#section-4.2.11
Clients MUST verify that the server's selected_identity is within the range supplied by the client
```